sdks/python/apache_beam/transforms/window_test.py

```diff
@@ -65,6 +76,44 @@ def process(self, element, window=core.DoFn.WindowParam):
 
 reify_windows = core.ParDo(ReifyWindowsFn())
 
+class TestCustomWindows(NonMergingWindowFn):
+  """A custom non merging window fn which assigns elements into interval windows
+  based on the element timestamps.
+  """
+
+  def __init__(self, first_window_end, second_window_end):
+    self.first_window_end = Timestamp.of(first_window_end)
+    self.second_window_end = Timestamp.of(second_window_end)
+
+  def assign(self, context):
+    timestamp = context.timestamp
+    if timestamp < self.first_window_end:
+      return [IntervalWindow(0, self.first_window_end)]
+    elif timestamp < self.second_window_end:
+      return [IntervalWindow(self.first_window_end, self.second_window_end)]
+    else:
+      return [IntervalWindow(self.second_window_end, timestamp)]
+
+  def get_window_coder(self):
+    return IntervalWindowCoder()
+
+  def to_runner_api_parameter(self, context):
```

Comment:
This custom window fn is added only for a test, to test a non-standard window fn. To fully implement it, I need to also implement to_runner_api_parameter and from_runner_api_parameter. Those require me to also change the standard_window_fns.proto and common_urns.py. Are these changes necessary? If not, is there a better way to do this? Thanks! @lukecwik @robertwb

Reply:
It doesn't seem like there is a good way for custom user window fns to return a python dict or some other structure and you currently need to translate the data you want to store in the window fn into some proto. Instead of defining a test windowfn proto, use one of the well known protobuf types such as [Struct](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/struct) or [BytesValue](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/bytes-value) and embed the information within those objects yourself instead of defining your own within StandardWindowFns.

Reply:
I made the custom window fn simpler in this version, by just overriding the assign() fn and the from_runner_api_parameter() fn. PTAL:) Thanks!

Reply:
With that change, I also need to separate the windows runner api tests to a separate file.